File: CMS/JS/FRAMEWORK/BussinessLogic/PermitsBO.js
﻿/**
 * @author Nam Chu
 * @created 02/0/15
 * 
 * @modified Nam Chu 08/07/15
 * -add some methods to work with ActionHandler
 * -add disable to Sel methods
 * -reformat and optimize code
 */
//TODO: Upd_Type, Upd_Status, Upd_Disable are NOT available
var PermitsBO = {
    Sel: function(disable) {
        var ret;
        if (disable == undefined) {
            jQuery.ajax({
                url: "/Action/ProcessBackendAction.ashx?ActionObject=Permits&action=Sel",
                type: "POST",
                dataType: "json",
                data: "",
                success: function(data, dataStatus) {
                    ret = data;
                },
                timeout: 30000,
                async: false,
                error: function(request, error) {}
            });
        } else {
            jQuery.ajax({
                url: "/Action/ProcessBackendAction.ashx?ActionObject=Permits&action=Sel&Disable=" + disable,
                type: "POST",
                dataType: "json",
                data: "",
                success: function(data, dataStatus) {
                    ret = data;
                },
                timeout: 30000,
                async: false,
                error: function(request, error) {}
            });
        }
        return ret;
    },

    Sel_ByID: function(id, disable) {
        var ret;
        if (disable == undefined) {
            jQuery.ajax({
                url: "/Action/ProcessBackendAction.ashx?ActionObject=Permits&action=Sel_ByID&ID=" + id,
                type: "POST",
                dataType: "json",
                data: "",
                success: function(data, dataStatus) {
                    ret = data;
                },
                timeout: 30000,
                async: false,
                error: function(request, error) {}
            });
        } else {
            jQuery.ajax({
                url: "/Action/ProcessBackendAction.ashx?ActionObject=Permits&action=Sel_ByID&ID=" + id + "&Disable=" + disable,
                type: "POST",
                dataType: "json",
                data: "",
                success: function(data, dataStatus) {
                    ret = data;
                },
                timeout: 30000,
                async: false,
                error: function(request, error) {}
            });
        }
        return ret;
    },

    Sel_ByType: function(type, disable) {
        var ret;
        if (disable == undefined) {
            jQuery.ajax({
                url: "/Action/ProcessBackendAction.ashx?ActionObject=Permits&action=Sel_ByType&Type=" + type,
                type: "POST",
                dataType: "json",
                data: "",
                success: function(data, dataStatus) {
                    ret = data;
                },
                timeout: 30000,
                async: false,
                error: function(request, error) {}
            });
        } else {
            jQuery.ajax({
                url: "/Action/ProcessBackendAction.ashx?ActionObject=Permits&action=Sel_ByType&Type=" + type + "&Disable=" + disable,
                type: "POST",
                dataType: "json",
                data: "",
                success: function(data, dataStatus) {
                    ret = data;
                },
                timeout: 30000,
                async: false,
                error: function(request, error) {}
            });
        }
        return ret;
    },

    Sel_ByStatus: function(status, disable) {
        var ret;
        if (disable == undefined) {
            jQuery.ajax({
                url: "/Action/ProcessBackendAction.ashx?ActionObject=Permits&action=Sel_ByStatus&Status=" + status,
                type: "POST",
                dataType: "json",
                data: "",
                success: function(data, dataStatus) {
                    ret = data;
                },
                timeout: 30000,
                async: false,
                error: function(request, error) {}
            });
        } else {
            jQuery.ajax({
                url: "/Action/ProcessBackendAction.ashx?ActionObject=Permits&action=Sel_ByStatus&Status=" + status + "&Disable=" + disable,
                type: "POST",
                dataType: "json",
                data: "",
                success: function(data, dataStatus) {
                    ret = data;
                },
                timeout: 30000,
                async: false,
                error: function(request, error) {}
            });
        }
        return ret;
    },

    Sel_ByType_ByStatus: function(type, status, disable) {
        var ret;
        if (disable == undefined) {
            jQuery.ajax({
                url: "/Action/ProcessBackendAction.ashx?ActionObject=Permits&action=Sel_ByType_ByStatus&Type=" + type + "&Status=" + status,
                type: "POST",
                dataType: "json",
                data: "",
                success: function(data, dataStatus) {
                    ret = data;
                },
                timeout: 30000,
                async: false,
                error: function(request, error) {}
            });
        } else {
            jQuery.ajax({
                url: "/Action/ProcessBackendAction.ashx?ActionObject=Permits&action=Sel_ByType_ByStatus&Type=" + type + "&Status=" + status + "&Disable=" + disable,
                type: "POST",
                dataType: "json",
                data: "",
                success: function(data, dataStatus) {
                    ret = data;
                },
                timeout: 30000,
                async: false,
                error: function(request, error) {}
            });
        }
        return ret;
    },

    Ins: function () {
        EnableLoading();
        jQuery.ajax({
            url: "/Action/ProcessBackendAction.ashx?ActionObject=Permits&action=Ins",
            type: "POST",
            dataType: "json",
            data: $("#frmIns_Permits").serialize(),

            success: function(data) {
                if (data.status == "success")
                    alert("Thêm Permits thành công.");
                else if (data.status != "success")
                    alert("Thêm Permits lỗi: <font style='font-size:9px'>" + data.message + "</font>");
                DisableLoading();
            },
            error: function(ex) {}
        });
    },

    Ins_ListPermits: function() {
        alert("Function Not Implemented");
    },

    Upd_ByID: function () {
        EnableLoading();
        jQuery.ajax({
            url: "/Action/ProcessBackendAction.ashx?ActionObject=Permits&action=Upd_ByID",
            type: "POST",
            dataType: "json",
            data: $("#frmUpd_Permits").serialize(),
            success: function(data) {
                if (data.status == "success")
                    alert("Update Permits thành công.");
                else if (data.status != "success")
                    alert("Update Permits lỗi: <font style='font-size:9px'>" + data.message + "</font>");
                DisableLoading();
            },
            error: function(ex) {}
        });
    },

    Upd_Type_ByID: function () {
        alert("Function Not Implemented");
    },

    Upd_Status_ByID: function () {
        alert("Function Not Implemented");
    },

    Upd_Disable_ByID: function () {
        alert("Function Not Implemented");
    },

    Del_ByID: function (id) {
        EnableLoading();
        jQuery.ajax({
            url: "/Action/ProcessBackendAction.ashx?ActionObject=Permits&action=Del_ByID&ID=" + id,
            type: "POST",
            dataType: "json",
            data: $("#frmUpd_Permits").serialize(),
            success: function(data) {
                if (data.status == "success")
                    alert("Delete Permits thành công.");
                else if (data.status != "success")
                    alert("Delete Permits lỗi: <font style='font-size:9px'>" + data.message + "</font>");
                DisableLoading();
            },
            error: function(ex) {}
        });
    },

    Del_ByListID: function() {
        alert("Function Not Implemented");
    }
}

File: BussinessLogic/PermitsBO.cs
using System;
using System.Collections.Generic;
using System.Data.Entity.Migrations;
using System.Linq;
using DataAccess;

// ReSharper disable once CheckNamespace

namespace BussinessLogic {
    /// <summary>
    /// PermitDetailsBO
    /// </summary>
    /// <author> Nam Chu </author>
    /// <created>10/06/15</created>
    /// <modified>01/07/15</modified>

    // ReSharper disable once InconsistentNaming
    public class PermitsBO {
        private readonly DatabaseDA _aDatabaseDa = new DatabaseDA();

        public List<vw_PermitsViewAll> GetAllInfoLogin_Ext_ByUsername(string username) {
            try {
                return _aDatabaseDa.vw_PermitsViewAll.Where(a => a.SystemUsers_Username == username).ToList();
            }
            catch (Exception ex) {
                throw new Exception(String.Format("SystemUsersDA_GetPermitViewAll_ByUsername: {0}", ex.Message));
            }
        }

        public List<vw_PermitsViewAll> GetAllInfoLogin_ByUsername(string userName) {
            try {
                return
                    _aDatabaseDa.vw_PermitsViewAll.Where(p => p.SystemUsers_Username == userName)
                        .Where(p => p.SystemUsers_Disable == false)
                        .ToList();
            }
            catch (Exception ex) {
                throw new Exception(String.Format("SystemUsersDA_GetAllInfoLogin_ByUsername: {0}", ex.Message));
            }
        }

        #region SELECT

        public List<Permits> Sel() {
            try {
                return _aDatabaseDa.Permits.ToList();
            }
            catch (Exception ex) {
                throw new Exception(String.Format("PermitsBO.Sel:{0}", ex.Message));
            }
        }

        public List<Permits> Sel(bool disable) {
            try {
                return _aDatabaseDa.Permits.Where(z => z.Disable == disable).ToList();
            }
            catch (Exception ex) {
                throw new Exception(String.Format("PermitsBO.Sel:{0}", ex.Message));
            }
        }

        public Permits Sel_ByID(Int32 id) {
            try {
                Permits aPermits = _aDatabaseDa.Permits.Where(b => b.ID == id).ToList()[0];
                return aPermits;
            }
            catch (Exception ex) {
                throw new Exception(String.Format("PermitsBO.Sel: {0}", ex.Message));
            }
        }

        public Permits Sel_ByID(Int32 id, bool disable) {
            try {
                Permits aPermits =
                    _aDatabaseDa.Permits.Where(b => b.ID == id).Where(b => b.Disable == disable).ToList()[0];
                return aPermits;
            }
            catch (Exception ex) {
                throw new Exception(String.Format("PermitsBO.Sel: {0}", ex.Message));
            }
        }

        public List<Permits> Sel_ByType(int type) {
            try {
                return _aDatabaseDa.Permits.Where(p => p.Type == type).ToList();
            }
            catch (Exception ex) {
                throw new Exception(String.Format("PermitsBO.Sel_ByType:{0}", ex.Message));
            }
        }

        public List<Permits> Sel_ByType(int type, bool disable) {
            try {
                return _aDatabaseDa.Permits.Where(p => p.Type == type).Where(z => z.Disable == disable).ToList();
            }
            catch (Exception ex) {
                throw new Exception(String.Format("PermitsBO.Sel_ByType:{0}", ex.Message));
            }
        }

        public List<Permits> Sel_ByStatus(int status) {
            try {
                return _aDatabaseDa.Permits.Where(p => p.Status == status).ToList();
            }
            catch (Exception ex) {
                throw new Exception(String.Format("PermitsBO.Sel_ByStatus:{0}", ex.Message));
            }
        }

        public List<Permits> Sel_ByStatus(int status, bool disable) {
            try {
                return _aDatabaseDa.Permits.Where(p => p.Status == status).Where(z => z.Disable == disable).ToList();
            }
            catch (Exception ex) {
                throw new Exception(String.Format("PermitsBO.Sel_ByStatus:{0}", ex.Message));
            }
        }

        public List<Permits> Sel_ByType_ByStatus(int type, int status) {
            try {
                return _aDatabaseDa.Permits.Where(c => c.Type == type).Where(p => p.Status == status).ToList();
            }
            catch (Exception ex) {
                throw new Exception(String.Format("PermitsBO.Sel_ByType_ByStatus:{0}", ex.Message));
            }
        }

        public List<Permits> Sel_ByType_ByStatus(int type, int status, bool disable) {
            try {
                return
                    _aDatabaseDa.Permits.Where(c => c.Type == type)
                        .Where(p => p.Status == status)
                        .Where(z => z.Disable == disable)
                        .ToList();
            }
            catch (Exception ex) {
                throw new Exception(String.Format("PermitsBO.Sel_ByType_ByStatus:{0}", ex.Message));
            }
        }

        #endregion

        #region DELETE

        public int Del_ByID(Int32 id) {
            try {
                List<Permits> aTemp = _aDatabaseDa.Permits.Where(p => p.ID == id).ToList();

                if (aTemp.Count > 0) {
                    _aDatabaseDa.Permits.Remove(aTemp[0]);
                    return _aDatabaseDa.SaveChanges();
                }
                else {
                    throw new Exception(String.Format("PermitsBO.Del_ByID: {0}", "Không tìm thấy Permits có ID = " + id));
                }
            }
            catch (Exception ex) {
                throw new Exception(String.Format("PermitsBO.Del: {0}", ex.Message));
            }
        }

        public int Del_ByListID(List<int> aListId) {
            try {
                List<Permits> aTemp =
                    _aDatabaseDa.Permits.Where(p => aListId.Contains(int.Parse(p.ID.ToString()))).ToList();
                if (aTemp.Count > 0) {
                    _aDatabaseDa.Permits.RemoveRange(aTemp);
                    return _aDatabaseDa.SaveChanges();
                }
                return 0;
            }
            catch (Exception ex) {
                throw new Exception(String.Format("PermitsBO.Del_By_ListID: {0}", ex.Message));
            }
        }

        #endregion

        #region UPDATE

        public int Upd(Permits aPermits) {
            try {
                _aDatabaseDa.Permits.AddOrUpdate(aPermits);
                return _aDatabaseDa.SaveChanges();
            }
            catch (Exception ex) {
                throw new Exception(String.Format("PermitsBO.Upd: {0}", ex.Message));
            }
        }

        public int Upd_Status(int id, int newStatus) {
            try {
                List<Permits> aListPermits = _aDatabaseDa.Permits.Where(b => b.ID == id).ToList();
                if (aListPermits.Count > 0) {
                    aListPermits[0].Status = newStatus;
                    _aDatabaseDa.Permits.AddOrUpdate(aListPermits[0]);
                    return _aDatabaseDa.SaveChanges();
                }
                return 0;
            }
            catch (Exception ex) {
                throw new Exception(String.Format("Upd_Status_ByID:{0}", ex.Message));
            }
        }

        public int Upd_Type(int id, int newType) {
            try {
                List<Permits> aListPermits = _aDatabaseDa.Permits.Where(b => b.ID == id).ToList();
                if (aListPermits.Count > 0) {
                    aListPermits[0].Type = newType;
                    _aDatabaseDa.Permits.AddOrUpdate(aListPermits[0]);
                    return _aDatabaseDa.SaveChanges();
                }
                return 0;
            }
            catch (Exception ex) {
                throw new Exception(String.Format("Upd_Type_ByID:{0}", ex.Message));
            }
        }

        public int Upd_Disable(int id, bool newDisable) {
            try {
                List<Permits> aListPermits = _aDatabaseDa.Permits.Where(b => b.ID == id).ToList();
                if (aListPermits.Count > 0) {
                    aListPermits[0].Disable = newDisable;
                    _aDatabaseDa.Permits.AddOrUpdate(aListPermits[0]);
                    return _aDatabaseDa.SaveChanges();
                }
                return 0;
            }
            catch (Exception ex) {
                throw new Exception(String.Format("Upd_Disable_ByID:{0}", ex.Message));
            }
        }

        public int Upd_ListPermit(List<Permits> aListPermits) {
            try {
                _aDatabaseDa.Permits.AddOrUpdate(aListPermits.ToArray());
                return _aDatabaseDa.SaveChanges();
            }
            catch (Exception ex) {
                throw new Exception(String.Format("PermitsBO.Ins: {0}", ex.Message));
            }
        }

        #endregion

        #region INSERT

        public int Ins(Permits aPermits) {
            try {
                _aDatabaseDa.Permits.AddOrUpdate(aPermits);
                return _aDatabaseDa.SaveChanges();
            }
            catch (Exception ex) {
                throw new Exception(String.Format("PermitsBO.Ins: {0}", ex.Message));
            }
        }

        public int Ins_ListPermits(ref List<Permits> aListPermits) {
            try {
                aListPermits = _aDatabaseDa.Permits.AddRange(aListPermits).ToList();
                return _aDatabaseDa.SaveChanges();
            }
            catch (Exception ex) {
                throw new Exception(String.Format("PermitsBO.Ins: {0}", ex.Message));
            }
        }

        #endregion
    }
}